### PR TITLE
Scheduler first fit 

### LINKED ISSUE #122057: Scheduler evaluates many nodes even if there are no scoring plugins defined
### What happened?

Even when no scoring plugins are defined, scheduler always attempts to find multiple nodes where the pod can run. This is unnecessary (and therefore inefficient) when we know ahead of time that there will be no difference between how those nodes are scored. 

### What did you expect to happen?

If we're going to pick a random node, we can just as well cut short the search after finding the first match.

### How can we reproduce it (as minimally and precisely as possible)?

Configure scheduler without any scoring plugins and see that mulitple nodes are evaluated anyway.

### Anything else we need to know?

/cc @ahg-g @alculquicondor 
/sig scheduling
/kind bug

### Kubernetes version

1.29 and older

### Cloud provider

N/A

### OS version

_No response_

### Install tools

_No response_

### Container runtime (CRI) and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by aleksandra-malinowska**
Resubmit of #122435, reverted  in #123290 due to increasing the chance of not finding an existing feasible node when extender filters are used. The issue itself is pre-existing and it's not clear whether it is something to be fixed (or how). This change is modified to exclude the cases when extenders are used.

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Skip looking for extra nodes if no scoring plugins and no extenders are defined. As there's no difference in priority between nodes when we don't score them and there's no risk that some won't pass extender filters, there's no point in searching for multiple fitting nodes.

#### Which issue(s) this PR fixes:

Fixes #122057

#### Does this PR introduce a user-facing change?

```release-note
Improves scheduler performance when no scoring plugins are defined.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig scheduling
/cc @ahg-g

### CODE REVIEW

File: pkg/scheduler/schedule_one.go [hunk 1]
```diff
@@ -751,14 +766,17 @@ func prioritizeNodes(
...
+	if len(sched.Extenders) != 0 && nodes != nil {
 		// allNodeExtendersScores has all extenders scores for all nodes.
 		// It is keyed with node name.
 		allNodeExtendersScores := make(map[string]*framework.NodePluginScores, len(nodes))
 		var mu sync.Mutex
 		var wg sync.WaitGroup
-		for i := range extenders {
-			if !extenders[i].IsInterested(pod) {
+		for i := range sched.Extenders {
+			if !sched.Extenders[i].IsInterested(pod) {
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:776]: Does this include Bind plugins?

File: pkg/scheduler/extender_test.go [hunk 2]
```diff
@@ -268,6 +270,30 @@ func TestSchedulerWithExtenders(t *testing.T) {
...
+					Predicates:   []tf.FitPredicate{tf.Node1PredicateExtender},
+				},
+			},
+			nodes: []string{"node1", "node2"},
+			expectedResult: ScheduleResult{
+				SuggestedHost:  "node1",
+				EvaluatedNodes: 2,
+				FeasibleNodes:  1,
+			},
+			name: "test 10",
```
Comment (alculquicondor) [inline pkg/scheduler/extender_test.go:295]: Use a name that reflects what the test is testing.

File: pkg/scheduler/schedule_one.go [hunk 3]
```diff
@@ -718,7 +733,7 @@ func prioritizeNodes(
 	logger := klog.FromContext(ctx)
 	// If no priority configs are provided, then all nodes will have a score of one.
 	// This is required to generate the priority list in the required format
-	if len(extenders) == 0 && !fwk.HasScorePlugins() {
+	if len(sched.Extenders) == 0 && !sched.hasScoring(fwk) {
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:736]: Shouldn't it be just this now?

```suggestion
	if !sched.hasScoring(fwk) {
```

File: pkg/scheduler/schedule_one.go [hunk 4]
```diff
@@ -751,14 +766,17 @@ func prioritizeNodes(
...
 		allNodeExtendersScores := make(map[string]*framework.NodePluginScores, len(nodes))
 		var mu sync.Mutex
 		var wg sync.WaitGroup
-		for i := range extenders {
-			if !extenders[i].IsInterested(pod) {
+		for i := range sched.Extenders {
+			if !sched.Extenders[i].IsInterested(pod) {
+				continue
+			}
+			if !sched.Extenders[i].IsPrioritizer() {
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:779]: Was the lack of this check a bug unrelated to the change?

If so, can you put it in a separate PR?

File: pkg/scheduler/schedule_one.go [hunk 5]
```diff
@@ -556,6 +569,9 @@ func (sched *Scheduler) findNodesThatPassFilters(
 	nodes []*framework.NodeInfo) ([]*framework.NodeInfo, error) {
 	numAllNodes := len(nodes)
 	numNodesToFind := sched.numFeasibleNodesToFind(fwk.PercentageOfNodesToScore(), int32(numAllNodes))
+	if len(sched.Extenders) == 0 && !sched.hasScoring(fwk) {
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:572]: This is not enough.

In an scenario where the scheduler only has a Bind extender, we could still set numNodesToFind to 1.

The check should be for `!hasScoring && !hasFilterExtenders`.

Comment (aleksandra-malinowska) [reply pkg/scheduler/schedule_one.go]: Done

File: pkg/scheduler/schedule_one.go (hunk 4 above)
Comment (aleksandra-malinowska) [reply pkg/scheduler/schedule_one.go]: Removed from this change

File: pkg/scheduler/schedule_one.go (hunk 3 above)
Comment (aleksandra-malinowska) [reply pkg/scheduler/schedule_one.go]: Fixed

File: pkg/scheduler/extender_test.go (hunk 2 above)
Comment (aleksandra-malinowska) [reply pkg/scheduler/extender_test.go]: Done

**Comment by AxeZhan** [comment]
Seems `gofmt` is failing, other than that lgtm.

File: pkg/scheduler/schedule_one_test.go [hunk 6]
```diff
@@ -2377,6 +2391,33 @@ func TestSchedulerSchedulePod(t *testing.T) {
...
+				tf.RegisterPreFilterPlugin(
+					"FakePreFilter",
+					tf.NewFakePreFilterPlugin("FakePreFilter", &framework.PreFilterResult{NodeNames: sets.New("node1", "node2")}, nil),
+				),
+				tf.RegisterBindPlugin(defaultbinder.Name, defaultbinder.New),
+			},
+			nodes:              []string{"node1", "node2", "node3"},
+			pod:                st.MakePod().Name("test-prefilter").UID("test-prefilter").Obj(),
+			wantNodes:          sets.New("node1", "node2"),
+			wantEvaluatedNodes: ptr.To[int32](2),
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one_test.go:2419]: why is it two in this case?

File: pkg/scheduler/testing/framework/fake_extender.go [hunk 7]
```diff
@@ -137,6 +137,30 @@ func (pl *node2PrioritizerPlugin) ScoreExtensions() framework.ScoreExtensions {
 	return nil
 }
 
+type equalPrioritizerPlugin struct{}
```
Comment (alculquicondor) [inline pkg/scheduler/testing/framework/fake_extender.go:140]: Can you use FakePreScoreAndScorePlugin instead of adding a new one?

File: pkg/scheduler/schedule_one_test.go [hunk 8]
```diff
@@ -1823,6 +1833,7 @@ func TestSchedulerSchedulePod(t *testing.T) {
 			registerPlugins: []tf.RegisterPluginFunc{
 				tf.RegisterQueueSortPlugin(queuesort.Name, queuesort.New),
 				tf.RegisterFilterPlugin("TrueFilter", tf.NewTrueFilterPlugin),
+				tf.RegisterScorePlugin("EqualPrioritizerPlugin", tf.NewEqualPrioritizerPlugin(), 20),
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one_test.go:1836]:
```suggestion
				tf.RegisterScorePlugin("EqualPrioritizerPlugin", tf.NewEqualPrioritizerPlugin(), 1),
```

1 makes it a bit more clear that the weight is irrelevant.

File: pkg/scheduler/schedule_one_test.go (hunk 6 above)
Comment (aleksandra-malinowska) [reply pkg/scheduler/schedule_one_test.go:2419]: #119779 changed the semantic of evaluated nodes (see also: https://github.com/kubernetes/kubernetes/pull/119779/files#r1439161728)

File: pkg/scheduler/testing/framework/fake_extender.go (hunk 7 above)
Comment (aleksandra-malinowska) [reply pkg/scheduler/testing/framework/fake_extender.go]: Changed NewEqualPrioritizerPlugin to be a wrapper creating a FakePreScoreAndScorePlugin.

File: pkg/scheduler/schedule_one_test.go (hunk 8 above)
Comment (aleksandra-malinowska) [reply pkg/scheduler/schedule_one_test.go]: Done

**Comment by aleksandra-malinowska** [comment]
@alculquicondor addressed latest comments in a new commit for easier review, I'll squash once the review is done.

File: pkg/scheduler/extender_test.go [hunk 9]
```diff
@@ -268,6 +270,30 @@ func TestSchedulerWithExtenders(t *testing.T) {
...
+					Predicates:   []tf.FitPredicate{tf.Node1PredicateExtender},
+				},
+			},
+			nodes: []string{"node1", "node2"},
+			expectedResult: ScheduleResult{
+				SuggestedHost:  "node1",
+				EvaluatedNodes: 2,
+				FeasibleNodes:  1,
+			},
+			name: "test 10 - no scoring, extender filters configured, multiple feasible nodes are evaluated",
```
Comment (alculquicondor) [inline pkg/scheduler/extender_test.go:295]: Let's add one test with no scoring and no extender filters, to ensure evaluated nodes is 1.

**Comment by alculquicondor** [comment]
/approve
@ahg-g anything to add?

**Comment by ahg-g** [comment]
> /approve @ahg-g anything to add?

closing my eyes and fingers crossed :)

/lgtm

File: pkg/scheduler/extender_test.go (hunk 9 above)
Comment (aleksandra-malinowska) [reply pkg/scheduler/extender_test.go:295]: Done
